src/solver/solvers/solver_ppc.h

```diff
@@ -132,151 +182,335 @@ namespace solver
+        auto level             = hierarchy->getPatchLevel(levelNumber);
+        auto dt                = newTime - currentTime;
 
-        auto level = hierarchy->getPatchLevel(levelNumber);
-        /*
-         * PREDICTOR 1
-         */
 
-        // loop on patches
-        // |
-        // -> faraday E , B, Bpred
-        // auto faraday = []() {};
-        // amr::visitLevel(*level, *hybridModel.resourcesManager, faraday, hybridModel);
 
-        VecFieldT& Bpred = electromagPred_.B;
-        fromCoarser.fillMagneticGhosts(Bpred, levelNumber, newTime);
+        predictor1_(*level, hybridModel, fromCoarser, currentTime, newTime);
 
 
-        // loop on patches
-        // |
-        // -> ampere Bpred, Jtot on interior + ghost
-        // -> ohm Bpred, ions.rho, Vepred1, PePred1, Jtot, Epred
+        average_(*level, hybridModel, fromCoarser, currentTime, newTime);
 
-        VecFieldT& Epred = electromagPred_.E;
-        fromCoarser.fillElectricGhosts(Epred, levelNumber, newTime);
-        // fromCoarser.getElectric(electromagPred_.E, fillTime, BooleanSelector<withTemporal>{},
-        //                        FillTypeSelector<FillType::GhostRegion>{});
+        moveIons_(*level, hybridState.ions, electromagAvg_, resourcesManager, fromCoarser,
+                  currentTime, newTime, core::UpdaterMode::moments_only);
 
-        // loop on patches
-        // |
-        // -> timeAverage E, Epred, Eavg
-        // -> timeAverage B, Bpred, Bavg
+        predictor2_(*level, hybridModel, fromCoarser, currentTime, newTime);
 
 
-        // loop on patches
-        // |
-        // -> moveIons_(Bavg, Eavg, ions, PredictorStep::predictor1);
+        average_(*level, hybridModel, fromCoarser, currentTime, newTime);
 
+        moveIons_(*level, hybridState.ions, electromagAvg_, resourcesManager, fromCoarser,
+                  currentTime, newTime, core::UpdaterMode::particles_and_moments);
 
-        // fill PRA and ghostsin purple region so that some of these
-        // particles may eventually enter the domain after being pushed
-        fromCoarser.fillIonGhostParticles(hybridState.ions, *level, newTime);
+        corrector_(*level, hybridModel, fromCoarser, currentTime, newTime);
 
-        // move all ions in:
-        //  - the domain
-        //  - the intra level ghost region
-        //  - coarse to fine ghost region
-        // accumulate those that are within the domain after being pushed (before pivot)
-        // see particle design code
 
-        // now some of the nodes close to the boundary of the patches (and level)
-        // are incomplete because they may have recieved contributions from particles outside
-        // their domain that would have entered the purple region during the push
-        // therefore we need to re-fill the purple region and accumulate that density
-        // this is done by calling a messenger to fill the moments.
+        // return newTime;
```

Comment:
will this be returned in future?

Reply:
not sure yet, I've let that, probably assessed in the multiple level PR